src/Draco.Compiler/Internal/Codegen/CSharpCodegen.cs

```diff
@@ -2,369 +2,415 @@
-        this.Indent1();
-        this.output.WriteLine($"{startLabel}:;");
-        var cond = this.VisitExpr(node.Condition.Value);
-        this.Indent2();
-        this.output.WriteLine($"if (!{cond}) goto {endLabel};");
-        this.VisitExpr(node.Expression);
-        this.Indent2();
-        this.output.WriteLine($"goto {startLabel};");
-        this.Indent1();
-        this.output.WriteLine($"{endLabel}:;");
-        return "default(Unit)";
+        var left = this.VisitExpr(node.Target);
+        var right = this.VisitExpr(node.Value);
+        var expr = MapAssignmentOperator(node.CompoundOperator, left, right);
+        this.WriteInstruction($"{expr};");
+        return left;
     }
 
-    public override string VisitStringExpr(Expr.String node)
+    public override string VisitStringExpr(Ast.Expr.String node)
     {
-        var result = this.AllocateRegister();
-        this.Indent2();
-        this.output.WriteLine($"dynamic {result} = new System.Text.StringBuilder();");
-        var firstInLine = true;
+        var builder = this.AllocateRegister();
+        this.WriteInstruction($"System.Text.StringBuilder {builder} = new System.Text.StringBuilder();");
         foreach (var part in node.Parts)
         {
-            if (part is StringPart.Interpolation i)
+            if (part is Ast.StringPart.Interpolation i)
             {
                 var subexpr = this.VisitExpr(i.Expression);
-                this.Indent2();
-                this.output.WriteLine($"{result}.Append({subexpr}.ToString());");
-                firstInLine = false;
+                this.WriteInstruction($"{builder}.Append({subexpr}.ToString());");
             }
             else
             {
-                var c = (StringPart.Content)part;
-                var text = c.Value.ValueText!;
-                if (firstInLine) text = text.Substring(c.Cutoff);
-                this.Indent2();
-                this.output.WriteLine($"{result}.Append(\"{StringUtils.Unescape(text)}\");");
-                firstInLine = c.Value.Type == TokenType.StringNewline;
+                var c = (Ast.StringPart.Content)part;
+                var text = c.Value;
+                text = text[c.Cutoff..];
+                this.WriteInstruction($"{builder}.Append(\"{StringUtils.Unescape(text)}\");");
             }
         }
-        return $"{result}.ToString()";
+        var result = this.DefineRegister(Type.String, expr: $"{builder}.ToString()");
+        return result!;
+    }
+
+    public override string? VisitLiteralExpr(Ast.Expr.Literal node) =>
+        node.Value?.ToString()?.ToLower();
+
+    public override string VisitReferenceExpr(Ast.Expr.Reference node) =>
+        this.AllocateName(node.Symbol);
+
+    private static string? MapUnaryOperator(ISymbol.IUnaryOperator op, string? sub)
+    {
+        if (sub is null) return null;
+
+        if (op == Intrinsics.Operators.Not_Bool) return $"!{sub}";
+
+        throw new NotImplementedException();
```

Comment:
Shouldn't we just return the operator?

Reply:
In general, the intrinsic can resolve to any "shape" of C# expression currently (look as mod and rem as an example), so to be uniform, no.